Documentation for plugins configuration. (#76)
mautic/developer-documentation-new: docs/plugins/integrations/authentication.rst

> **************************

Reviewer: ```suggestion
```

Integration authentication
**************************

> .. contents:: Table of contents

Reviewer: ```suggestion
```
I don't think that we need to add a ToC on every page - it takes a lot of space and they are navigable from the menu.

The IntegrationsBundle provides factories and helpers to create Guzzle Client classes for common authentication protocols.

----------

Registering the Integration for authentication
##############################################

If the Integration requires the User to authenticate through the web (OAuth2 three legged), the Integration needs to tag a service with ``mautic.auth_integration`` to handle the authentication process (redirecting to login, request the access token, etc).

> This service needs to implement::

Reviewer: ```suggestion
This service needs to implement:
```

I think only one colon is needed here?

>     \Mautic\IntegrationsBundle\Integration\Interfaces\AuthenticationInterface.

Reviewer: Is the ending full stop required here?

.. code-block:: php

    <?php
    return [
        // ...
        'services' => [
            // ...
            'integrations' => [
                // ...
                'helloworld.integration.authentication' => [
                    'class' => \MauticPlugin\HelloWorldBundle\Integration\Support\AuthSupport::class,
                    'tags'  => [
                        'mautic.auth_integration',
                    ],
                ],
                // ...
            ],
            // ...
        ],
        // ...
    ];

> 

Reviewer: Is it appropriate to use the RST markup to explain in more depth some of these examples - like how it is done on these pages:

https://mautic-developer.readthedocs.io/en/latest/plugins/config.html 

What do you think?

The ``AuthSupport`` class must implement ``\Mautic\IntegrationsBundle\Integration\Interfaces\AuthenticationInterface``.

.. code-block:: php

    <?php
    namespace MauticPlugin\HelloWorldBundle\Integration\Support;

    use MauticPlugin\HelloWorldBundle\Connection\Client;
    use MauticPlugin\HelloWorldBundle\Integration\HelloWorldIntegration;

>     use MauticPlugin\IntegrationsBundle\Integration\ConfigurationTrait;
>     use MauticPlugin\IntegrationsBundle\Integration\Interfaces\AuthenticationInterface;

Reviewer 2: This uses the wrong namespace 😊 

```suggestion
    use Mautic\IntegrationsBundle\Integration\ConfigurationTrait;
    use Mautic\IntegrationsBundle\Integration\Interfaces\AuthenticationInterface;
```

    use Symfony\Component\HttpFoundation\Request;

    class AuthSupport implements AuthenticationInterface
    {
        use ConfigurationTrait;

>         /**
>          * @var Client
>          */
>         private $client;

Reviewer 2: Let's use some PHP 7.4+ syntax here

```suggestion
        private Client $client;
```

        public function getName(): string
        {
            return HelloWorldIntegration::NAME;
        }

        public function getDisplayName(): string
        {
            return 'Hello World';
        }

        /**
         * Returns true if the integration has already been authorized with the third party service.
         *
         * @return bool
         */
        public function isAuthenticated(): bool
        {
            $apiKeys = $this->getIntegrationConfiguration()->getApiKeys();

            return !empty($apiKeys['access_token']) && !empty($apiKeys['refresh_token']);
        }

        /**
         * Authenticate and obtain the access token
         *
         * @param Request $request
         *
         * @return string
         */
        public function authenticateIntegration(Request $request): string
        {
            $code = $request->query->get('code');

            $this->client->authenticate($code);

            return 'Success!';
        }
    }

Authentication providers
************************

The Integration bundle comes with a number of popular authentication protocols available to use as Guzzle clients. New ones should implement::

>     \Mautic\IntegrationsBundle\Auth\Provider\AuthProviderInterface.

Reviewer 2: ```suggestion
    \Mautic\IntegrationsBundle\Auth\Provider\AuthProviderInterface
```

> **The examples below use anonymous classes. Use Object Oriented Programming with services and factories to generate credential, configuration, and client classes.**

Reviewer: Does this need to be bold, or would it be better as a note or a tip?

The best way to get configuration values such as username, password, consumer key, consumer secret, and so forth is by using the ``mautic.integrations.helper`` ``(\Mautic\IntegrationsBundle\Helper\IntegrationsHelper)`` service to leverage the configuration stored in the ``Integration`` entity's API keys.

.. code-block:: php

    <?php
    use MauticPlugin\HelloWorldBundle\Integration\HelloWorldIntegration;

>     use MauticPlugin\IntegrationsBundle\Helper\IntegrationsHelper;

Reviewer 2: ```suggestion
    use Mautic\IntegrationsBundle\Helper\IntegrationsHelper;
```

    /** @var $integrationsHelper IntegrationsHelper */
    $integration = $integrationsHelper->getIntegration(HelloWorldIntegration::NAME);

    $configuration = $integration->getIntegrationConfiguration();
    $apiKeys       = $configuration->getApiKeys();

    $username = $apiKeys['username'] ?? null;
    $password = $apiKeys['password'] ?? null;

    //...

API key
=======

Use the ``mautic.integrations.auth_provider.api_key`` service (``\Mautic\IntegrationsBundle\Auth\Provider\ApiKey\HttpFactory``) to obtain a ``GuzzleHttp\ClientInterface`` that uses an API key for all requests. Out of the box, the factory supports a parameter API key or a header API key.

Parameter based API key
-----------------------

To use the parameter based API key, create a credentials class that implements::

>     \Mautic\IntegrationsBundle\Auth\Provider\ApiKey\Credentials\ParameterCredentialsInterface.

Reviewer 2: ```suggestion
    \Mautic\IntegrationsBundle\Auth\Provider\ApiKey\Credentials\ParameterCredentialsInterface
```

.. code-block:: php

    <?php
    use MauticPlugin\HelloWorldBundle\Integration\HelloWorldIntegration;

>     use MauticPlugin\IntegrationsBundle\Auth\Provider\ApiKey\Credentials\ParameterCredentialsInterface;
>     use MauticPlugin\IntegrationsBundle\Auth\Provider\ApiKey\HttpFactory;
>     use MauticPlugin\IntegrationsBundle\Helper\IntegrationsHelper;

Reviewer 2: ```suggestion
    use Mautic\IntegrationsBundle\Auth\Provider\ApiKey\Credentials\ParameterCredentialsInterface;
    use Mautic\IntegrationsBundle\Auth\Provider\ApiKey\HttpFactory;
    use Mautic\IntegrationsBundle\Helper\IntegrationsHelper;
```

    /** @var $integrationsHelper IntegrationsHelper */
    $integration = $integrationsHelper->getIntegration(HelloWorldIntegration::NAME);

    $apiKeys = $integration->getIntegrationConfiguration()->getApiKeys();

    $credentials = new class($apiKeys['api_key']) implements ParameterCredentialsInterface {
        private $key;

        public function __construct(string $key)
        {
            $this->key = $key;
        }

        public function getKeyName(): string
        {
            return 'apikey';
        }

        public function getApiKey(): string
        {
            return $this->key;
        }
    };

    /** @var $factory HttpFactory */
    $client   = $factory->getClient($credentials);
    $response = $client->get('https://example.com/api/fetch');

Header based API key
--------------------

.. code-block:: php

    <?php
    use MauticPlugin\HelloWorldBundle\Integration\HelloWorldIntegration;

>     use MauticPlugin\IntegrationsBundle\Auth\Provider\ApiKey\Credentials\HeaderCredentialsInterface;
>     use MauticPlugin\IntegrationsBundle\Auth\Provider\ApiKey\HttpFactory;
>     use MauticPlugin\IntegrationsBundle\Helper\IntegrationsHelper;

Reviewer 2: ```suggestion
    use Mautic\IntegrationsBundle\Auth\Provider\ApiKey\Credentials\HeaderCredentialsInterface;
    use Mautic\IntegrationsBundle\Auth\Provider\ApiKey\HttpFactory;
    use Mautic\IntegrationsBundle\Helper\IntegrationsHelper;
```

    /** @var $integrationsHelper IntegrationsHelper */
    $integration = $integrationsHelper->getIntegration(HelloWorldIntegration::NAME);

    $apiKeys = $integration->getIntegrationConfiguration()->getApiKeys();

    $credentials = new class($apiKeys['api_key']) implements HeaderCredentialsInterface {
        private $key;

        public function __construct(string $key)
        {
            $this->key = $key;
        }

        public function getKeyName(): string
        {
            return 'X-API-KEY';
        }

        public function getApiKey(): string
        {
            return $this->key;
        }
    };

    /** @var $factory HttpFactory */
    $client   = $factory->getClient($credentials);
    $response = $client->get('https://example.com/api/fetch');

Basic auth
==========

Use the ``mautic.integrations.auth_provider.basic_auth`` service (``\Mautic\IntegrationsBundle\Auth\Provider\BasicAuth\HttpFactory``) to obtain a ``GuzzleHttp\ClientInterface`` that uses basic auth for all requests.

.. code-block:: php

    <?php
    use MauticPlugin\HelloWorldBundle\Integration\HelloWorldIntegration;

>     use MauticPlugin\IntegrationsBundle\Helper\IntegrationsHelper;
>     use MauticPlugin\IntegrationsBundle\Auth\Provider\BasicAuth\HttpFactory;
>     use MauticPlugin\IntegrationsBundle\Auth\Provider\BasicAuth\CredentialsInterface;

Reviewer 2: ```suggestion
    use Mautic\IntegrationsBundle\Helper\IntegrationsHelper;
    use Mautic\IntegrationsBundle\Auth\Provider\BasicAuth\HttpFactory;
    use Mautic\IntegrationsBundle\Auth\Provider\BasicAuth\CredentialsInterface;
```

    $configuration = $integration->getIntegrationConfiguration();
    $apiKeys       = $configuration->getApiKeys();

    $credentials = new class($apiKeys['username'], $apiKeys['password']) implements CredentialsInterface {
        private $username;
        private $password;

        public function __construct(string $username, string $password)
        {
            $this->username = $username;
            $this->password = $password;
        }

        public function getUsername(): string
        {
            return $this->username;
        }

        public function getPassword(): string
        {
            return $this->password;
        }
    };

    /** @var $factory HttpFactory */
    $client   = $factory->getClient($credentials);
    $response = $client->get('https://example.com/api/fetch');

OAuth1a
=======

OAuth1a three legged
--------------------

This has not been implemented yet.

OAuth1a two legged
------------------

OAuth1a two legged does not require a User to login as would three legged.

.. code-block:: php

    <?php
    use MauticPlugin\HelloWorldBundle\Integration\HelloWorldIntegration;

>     use MauticPlugin\IntegrationsBundle\Helper\IntegrationsHelper;
>     use MauticPlugin\IntegrationsBundle\Auth\Provider\OAuth1aTwoLegged\HttpFactory;
>     use MauticPlugin\IntegrationsBundle\Auth\Provider\OAuth1aTwoLegged\CredentialsInterface;

Reviewer 2: ```suggestion
    use Mautic\IntegrationsBundle\Helper\IntegrationsHelper;
    use Mautic\IntegrationsBundle\Auth\Provider\OAuth1aTwoLegged\HttpFactory;
    use Mautic\IntegrationsBundle\Auth\Provider\OAuth1aTwoLegged\CredentialsInterface;
```

        {
            return $this->consumerSecret;
        }

        /**
         * Not used in this example. Tsk tsk for breaking the interface segregation principle
         *
         * @return string|null
         */
        public function getToken(): ?string
        {
            return null;
        }

        /**
         * Not used in this example. Tsk tsk for breaking the interface segregation principle
         *
         * @return string|null
         */
        public function getTokenSecret(): ?string
        {
            return null;
        }
    };

    /** @var $factory HttpFactory */
    $client   = $factory->getClient($credentials);
    $response = $client->get('https://example.com/api/fetch');

OAuth2
======

Use the OAuth2 factory according to the grant type required. ``\Mautic\IntegrationsBundle\Auth\Provider\Oauth2ThreeLegged\HttpFactory`` supports ``code`` and ``refresh_token`` grant types. ``\Mautic\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\HttpFactory`` supports ``client_credentials`` and ``password``.

The OAuth2 factories leverages :xref:`Guzzle Oauth2 Subscriber` as a middleware.

Client configuration
--------------------

> Both OAuth2 factories leverage the ``\Mautic\IntegrationsBundle\Auth\Provider\AuthConfigInterface`` object to manage things such as configuring the signer (basic auth, post form data, custom), token factory, token persistence, and token signer (bearer auth, basic auth, query string, custom). Use the appropriate interfaces as required for the use case (see the interfaces in ``plugins/IntegrationsBundle/Auth/Support/Oauth2/ConfigAccess``).

Reviewer 2: ```suggestion
Both OAuth2 factories leverage the ``\Mautic\IntegrationsBundle\Auth\Provider\AuthConfigInterface`` object to manage things such as configuring the signer (basic auth, post form data, custom), token factory, token persistence, and token signer (bearer auth, basic auth, query string, custom). Use the appropriate interfaces as required for the use case (see the interfaces in ``app/bundles/IntegrationsBundle/Auth/Support/Oauth2/ConfigAccess``).
```

See :xref:`Guzzle Oauth2 Subscriber` for additional details on configuring the credentials and token signers or creating custom token persistence and factories.

Token persistence
-----------------

For most use cases, a token persistence service to fetch and store the access tokens generated by using refresh tokens, etc are required. The IntegrationBundle provides one that natively uses the ``\Mautic\PluginBundle\Entity\Integration`` entity's API keys. Anything stored through the service is automatically encrypted.

Use the ``mautic.integrations.auth_provider.token_persistence_factory`` service (``\Mautic\IntegrationsBundle\Auth\Support\Oauth2\Token\TokenPersistenceFactory``) to generate a ``TokenFactoryInterface`` to be returned by the ``\Mautic\IntegrationsBundle\Auth\Support\Oauth2\ConfigAccess\ConfigTokenFactoryInterface`` interface.
 
.. code-block:: php

    <?php
    use kamermans\OAuth2\Persistence\TokenPersistenceInterface;
    use MauticPlugin\HelloWorldBundle\Integration\HelloWorldIntegration;

>     use MauticPlugin\IntegrationsBundle\Auth\Support\Oauth2\ConfigAccess\ConfigTokenPersistenceInterface;
>     use MauticPlugin\IntegrationsBundle\Auth\Support\Oauth2\Token\TokenPersistenceFactory;
>     use MauticPlugin\IntegrationsBundle\Helper\IntegrationsHelper;

Reviewer 2: ```suggestion
    use Mautic\IntegrationsBundle\Auth\Support\Oauth2\ConfigAccess\ConfigTokenPersistenceInterface;
    use Mautic\IntegrationsBundle\Auth\Support\Oauth2\Token\TokenPersistenceFactory;
    use Mautic\IntegrationsBundle\Helper\IntegrationsHelper;
```

        public function __construct(TokenPersistenceInterface$tokenPersistence)
        {
            $this->tokenPersistence = $tokenPersistence;
        }

        public function getTokenPersistence(): TokenPersistenceInterface
        {
            return $this->tokenPersistence;
        }
    };

The token persistence service automatically manages ``access_token``, ``refresh_token``, and ``expires_at`` from the authentication process and stores them in the ``Integration`` entity's API keys array.

Token factory
-------------

In some cases, the third party service may return additional values that are not traditionally part of the OAuth2 spec and these values are required for further communication with the API service. In this case, the integration bundle's ``\Mautic\IntegrationsBundle\Auth\Support\Oauth2\Token\IntegrationTokenFactory`` can be used to capture those extra values and store them in the ``Integration`` entity's API keys array.

The ``IntegrationTokenFactory`` can then be returned in a ``\Mautic\IntegrationsBundle\Auth\Support\Oauth2\ConfigAccess\ConfigTokenFactoryInterface`` when configuring the ``Client``.

.. code-block:: php

    <?php

>     use MauticPlugin\IntegrationsBundle\Auth\Support\Oauth2\ConfigAccess\ConfigTokenFactoryInterface;
>     use MauticPlugin\IntegrationsBundle\Auth\Support\Oauth2\Token\IntegrationTokenFactory;
>     use MauticPlugin\IntegrationsBundle\Auth\Support\Oauth2\Token\TokenFactoryInterface;

Reviewer 2: ```suggestion
    use Mautic\IntegrationsBundle\Auth\Support\Oauth2\ConfigAccess\ConfigTokenFactoryInterface;
    use Mautic\IntegrationsBundle\Auth\Support\Oauth2\Token\IntegrationTokenFactory;
    use Mautic\IntegrationsBundle\Auth\Support\Oauth2\Token\TokenFactoryInterface;
```

    $tokenFactory = new IntegrationTokenFactory(['something_extra']);

    $config = new class($tokenFactory) implements ConfigTokenFactoryInterface {
        private $tokenFactory;

        public function __construct(TokenFactoryInterface $tokenFactory)
        {
            $this->tokenFactory = $tokenFactory;
        }

        public function getTokenFactory(): TokenFactoryInterface
        {
            return $this->tokenFactory;
        }
    };

OAuth2 two legged
=================

Password grant
--------------

> Below is an example of the password grant for a service that uses a scope (optional interface). The use of the token persistence is assuming the access token is valid for a period of time (that is an hour).

Reviewer: Do you mean that the default is an hour? If so let's be clear.

.. code-block:: php

    <?php
    use kamermans\OAuth2\Persistence\TokenPersistenceInterface;
    use MauticPlugin\HelloWorldBundle\Integration\HelloWorldIntegration;

>     use MauticPlugin\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\Credentials\PasswordCredentialsGrantInterface;
>     use MauticPlugin\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\Credentials\ScopeInterface;
>     use MauticPlugin\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\HttpFactory;
>     use MauticPlugin\IntegrationsBundle\Auth\Support\Oauth2\ConfigAccess\ConfigTokenPersistenceInterface;
>     use MauticPlugin\IntegrationsBundle\Helper\IntegrationsHelper;

Reviewer 2: ```suggestion
    use Mautic\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\Credentials\PasswordCredentialsGrantInterface;
    use Mautic\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\Credentials\ScopeInterface;
    use Mautic\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\HttpFactory;
    use Mautic\IntegrationsBundle\Auth\Support\Oauth2\ConfigAccess\ConfigTokenPersistenceInterface;
    use Mautic\IntegrationsBundle\Helper\IntegrationsHelper;
```

        {
            return $this->clientSecret;
        }

        public function getPassword(): ?string
        {
            return $this->password;
        }

        public function getUsername(): ?string
        {
            return $this->username;
        }

        public function getScope(): ?string
        {
            return $this->scope;
        }
    };

    /** @var $tokenPersistenceFactory TokenPersistenceFactory */
    $tokenPersistence = $tokenPersistenceFactory->create($integration);
    $config           = new class($tokenPersistence) implements ConfigTokenPersistenceInterface {
        private $tokenPersistence;

        public function __construct(TokenPersistenceInterface$tokenPersistence)
        {
            $this->tokenPersistence = $tokenPersistence;
        }

        public function getTokenPersistence(): TokenPersistenceInterface
        {
            return $this->tokenPersistence;
        }
    };

    /** @var $factory HttpFactory */
    $client   = $factory->getClient($credentials, $config);
    $response = $client->get('https://example.com/api/fetch');

Client credentials grant
------------------------

> Below is an example of the client credentials grant for a service that uses a scope (optional interface). The use of the token persistence is assuming the access token is valid for a period of time (that is an hour).

Reviewer: As above - if this is the default let's be clear that is what we are meaning.

.. code-block:: php

    <?php
    use kamermans\OAuth2\Persistence\TokenPersistenceInterface;
    use MauticPlugin\HelloWorldBundle\Integration\HelloWorldIntegration;

>     use MauticPlugin\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\Credentials\ClientCredentialsGrantInterface;
>     use MauticPlugin\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\Credentials\ScopeInterface;
>     use MauticPlugin\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\HttpFactory;
>     use MauticPlugin\IntegrationsBundle\Auth\Support\Oauth2\ConfigAccess\ConfigTokenPersistenceInterface;
>     use MauticPlugin\IntegrationsBundle\Helper\IntegrationsHelper;

Reviewer 2: ```suggestion
    use Mautic\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\Credentials\ClientCredentialsGrantInterface;
    use Mautic\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\Credentials\ScopeInterface;
    use Mautic\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\HttpFactory;
    use Mautic\IntegrationsBundle\Auth\Support\Oauth2\ConfigAccess\ConfigTokenPersistenceInterface;
    use Mautic\IntegrationsBundle\Helper\IntegrationsHelper;
```

        {
            return $this->authorizeUrl;
        }

        public function getClientId(): ?string
        {
            return $this->clientId;
        }

        public function getClientSecret(): ?string
        {
            return $this->clientSecret;
        }

        public function getScope(): ?string
        {
            return $this->scope;
        }
    };

    /** @var $tokenPersistenceFactory TokenPersistenceFactory */
    $tokenPersistence = $tokenPersistenceFactory->create($integration);
    $config           = new class($tokenPersistence) implements ConfigTokenPersistenceInterface {
        private $tokenPersistence;

        public function __construct(TokenPersistenceInterface$tokenPersistence)
        {
            $this->tokenPersistence = $tokenPersistence;
        }

        public function getTokenPersistence(): TokenPersistenceInterface
        {
            return $this->tokenPersistence;
        }
    };

    /** @var $factory HttpFactory */
    $client   = $factory->getClient($credentials, $config);
    $response = $client->get('https://example.com/api/fetch');

OAuth2 three legged
===================

> Three legged OAuth2 with the code grant is the most complex to implement because it involves redirecting the user to the third party service to authenticate then sent back to Mautic to initiate the access token process using a code returned in the request.

Reviewer: ```suggestion
Three legged OAuth2 with the code grant is the most complex to implement because it involves redirecting the User to the third party service to authenticate then sent back to Mautic to initiate the access token process using a code returned in the request.
```

> The first step is to register the integration as a :ref:`\\Mautic\\IntegrationsBundle\\Integration\\Interfaces\\AuthenticationInterface<Registering the Integration for Authentication>`. The ``authenticateIntegration()`` method initiates the access token process using the ``code`` returned in the request after the user logs into the third-party service. The Integration bundle provides a route that can use as the redirect or callback URIs through the named route ``mautic_integration_public_callback`` that requires a ``integration`` parameter. This redirect URI can display in the UI by using :xref:`ConfigFormCallbackInterface`. This route is to find the integration by name from the ``AuthIntegrationsHelper``and then execute its ``authenticateIntegration()``.

Reviewer: ```suggestion
The first step is to register the Integration as a :ref:`\\Mautic\\IntegrationsBundle\\Integration\\Interfaces\\AuthenticationInterface<Registering the Integration for Authentication>`. The ``authenticateIntegration()`` method initiates the access token process using the ``code`` returned in the request after the user logs into the third-party service. The Integration bundle provides a route that can use as the redirect or callback URIs through the named route ``mautic_integration_public_callback`` that requires a ``integration`` parameter. This redirect URI can display in the UI by using :xref:`ConfigFormCallbackInterface`. This route is to find the integration by name from the ``AuthIntegrationsHelper`` and then execute its ``authenticateIntegration()``.
```

.. code-block:: php

    <?php
    namespace MauticPlugin\HelloWorldBundle\Integration\Support;

    use GuzzleHttp\ClientInterface;

>     use MauticPlugin\IntegrationsBundle\Integration\Interfaces\AuthenticationInterface;

Reviewer 2: ```suggestion
    use Mautic\IntegrationsBundle\Integration\Interfaces\AuthenticationInterface;
```

    use Symfony\Component\HttpFoundation\Response;

    class AuthSupport implements AuthenticationInterface {
        /**
         * @var ClientInterface
         */
        private $client;

        // ...

        public function authenticateIntegration(Request $request): Response
        {
            $code = $request->query->get('code');

            $this->client->authenticate($code);

            return new Response('OK!');
        }
    }

The trick here is that the ``Client``'s ``authenticate`` method configures a ``ClientInterface`` and then calls to any valid API URL (*this is required*). The middleware initiates the access token process by making a call and storing it in the ``Integration`` entity's API keys through :ref:`TokenPersistenceFactory<Token Persistence>`. The URL is recommended to be something simple, like a version check or fetching info for the authenticated User.

Here is an example of a client, assuming that the User has already logged in and the code is in the request.

.. code-block:: php

    <?php
    use kamermans\OAuth2\Persistence\TokenPersistenceInterface;
    use MauticPlugin\HelloWorldBundle\Integration\HelloWorldIntegration;

>     use MauticPlugin\IntegrationsBundle\Auth\Provider\Oauth2ThreeLegged\Credentials\CodeInterface;
>     use MauticPlugin\IntegrationsBundle\Auth\Provider\Oauth2ThreeLegged\Credentials\CredentialsInterface;
>     use MauticPlugin\IntegrationsBundle\Auth\Provider\Oauth2ThreeLegged\Credentials\RedirectUriInterface;
>     use MauticPlugin\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\Credentials\ScopeInterface;
>     use MauticPlugin\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\HttpFactory;
>     use MauticPlugin\IntegrationsBundle\Auth\Support\Oauth2\ConfigAccess\ConfigTokenPersistenceInterface;
>     use MauticPlugin\IntegrationsBundle\Helper\IntegrationsHelper;

Reviewer 2: ```suggestion
    use Mautic\IntegrationsBundle\Auth\Provider\Oauth2ThreeLegged\Credentials\CodeInterface;
    use Mautic\IntegrationsBundle\Auth\Provider\Oauth2ThreeLegged\Credentials\CredentialsInterface;
    use Mautic\IntegrationsBundle\Auth\Provider\Oauth2ThreeLegged\Credentials\RedirectUriInterface;
    use Mautic\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\Credentials\ScopeInterface;
    use Mautic\IntegrationsBundle\Auth\Provider\Oauth2TwoLegged\HttpFactory;
    use Mautic\IntegrationsBundle\Auth\Support\Oauth2\ConfigAccess\ConfigTokenPersistenceInterface;
    use Mautic\IntegrationsBundle\Helper\IntegrationsHelper;
```
